KAFKA-9434: automated protocol for alterReplicaLogDirs

tombentley opened the pull request:
Change the alterReplicaLogDirs request and response to use the generated protocol messages.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tombentley committed:
wip, need to fix KafkaApis/ReplicaManager

tombentley committed:
wip

tombentley committed:
Minor improvements

tombentley commented:
@mimaison would you be able to review this one?

tombentley commented:
Rebased for conflict.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -18,140 +18,87 @@
 package org.apache.kafka.common.requests;
```
Leaving this comment here because there is not other place. While we are refactoring them, could we ensure that both the request and the response are covered in `org.apache.kafka.common.requests.RequestResponseTest`? I had a very quick look and it seems that they are not.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -18,140 +18,87 @@
+                return objectStream;
+            }).forEach(tp -> result.put(tp, alter.path()));
+        }
+        return result;
```
The `flatMap` seems wasteful here. Couldn't we use nested `forEach`s and add the `TopicPartition` and the `path` to the `HashMap`?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -18,140 +18,87 @@
+        AlterReplicaLogDirsResponseData data = new AlterReplicaLogDirsResponseData();
+
+        data.setResults(this.data.dirs().stream().flatMap(x -> {
+            Stream<AlterReplicaLogDirTopicResult> y = x.topics().stream().map(topic -> {
```
Could we please use better names than `x` and `y`?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -18,140 +18,87 @@
+                                            .setPartitionIndex(partitionId);
+                                }).collect(Collectors.toList()));
+            });
+            return y;
```
nit: What about putting this `return` at L77 directly?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -18,140 +18,87 @@
+                                    return new AlterReplicaLogDirsResponseData.AlterReplicaLogDirPartitionResult()
+                                            .setErrorCode(Errors.forException(e).code())
+                                            .setPartitionIndex(partitionId);
+                                }).collect(Collectors.toList()));
```
nit: `{ }` and `return` are not necessary in one-line lambda bodies. I would remove them here and in the other places to be consistent with rest of the code base.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsResponse.java:
```diff
@@ -17,122 +17,59 @@
-        return errorCounts(responses.values());
+        Map<Errors, Integer> errorCounts = new HashMap<>();
+        for (AlterReplicaLogDirTopicResult x : data.results()) {
+            for (AlterReplicaLogDirsResponseData.AlterReplicaLogDirPartitionResult y : x.partitions()) {
```
nit: Could we rename `x` and `y` here as well?

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2555,13 +2555,26 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleAlterReplicaLogDirsRequest(request: RequestChannel.Request): Unit = {
     val alterReplicaDirsRequest = request.body[AlterReplicaLogDirsRequest]
-    val responseMap = {
+    def sendResponseCallback(requestThrottleMs: Int): AlterReplicaLogDirsResponse = {
```
I would prefer if we could keep the previous structure where `authorize` and `replicaManager.alterReplicaLogDirs` were called before the `sendResponseMaybeThrottle`. Having everything in the callback, coupled with the response construction, makes the code hard to read.

tombentley replied:
I changed it because both branches of the `if authorize() ...` require the `requestThrottleMs` to set on the `AlterReplicaLogDirsResponseData`. I can set the throttle outside the `if` with something like

```scala
sendResponseMaybeThrottle(request, requestThrottleMs => {
      response.data().setThrottleTimeMs(requestThrottleMs)
      response
    })
```

It would then be more natural to call the `alterReplicaDirsRequest.getErrorResponse(Errors.CLUSTER_AUTHORIZATION_FAILED.exception)` which doesn't take a throttle (since we'll be setting it later), but that requires overriding that method so that the return type is not `AbstractResponse`. 

It's a bit unsatisfying, but did you have a better idea?

dajac replied:
I was thinking about something like this:
```
  def handleAlterReplicaLogDirsRequest(request: RequestChannel.Request): Unit = {
    val alterReplicaDirsRequest = request.body[AlterReplicaLogDirsRequest]
    if (authorize(request.context, ALTER, CLUSTER, CLUSTER_NAME)) {
      val result = replicaManager.alterReplicaLogDirs(alterReplicaDirsRequest.partitionDirs.asScala)
      sendResponseMaybeThrottle(request, requestThrottleMs =>
        new AlterReplicaLogDirsResponse(new AlterReplicaLogDirsResponseData()
          .setResults(result.groupBy(_._1.topic).map {
            case (topic, errors) => new AlterReplicaLogDirsResponseData.AlterReplicaLogDirTopicResult()
              .setTopicName(topic)
              .setPartitions(errors.map { 
                case (tp, error) => new AlterReplicaLogDirsResponseData.AlterReplicaLogDirPartitionResult()
                  .setPartitionIndex(tp.partition)
                  .setErrorCode(error.code)
              }.toList.asJava)
            }.toList.asJava)
          .setThrottleTimeMs(requestThrottleMs))
    } else {
      sendResponseMaybeThrottle(request, requestThrottleMs =>
        alterReplicaDirsRequest.getErrorResponse(requestThrottleMs, Errors.CLUSTER_AUTHORIZATION_FAILED.exception))
    }
  }
```

I find it a bit easier to read and it is also a bit more consistent with the implementation of the other handlers.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2555,13 +2555,26 @@ class KafkaApis(val requestChannel: RequestChannel,
 
   def handleAlterReplicaLogDirsRequest(request: RequestChannel.Request): Unit = {
```
While we are refactoring this, could we add a unit test in `KafkaApisTest`?

dajac commented on core/src/test/scala/unit/kafka/server/AlterReplicaLogDirsRequestTest.scala:
```diff
@@ -104,13 +112,26 @@ class AlterReplicaLogDirsRequestTest extends BaseRequestTest {
 
   private def sendAlterReplicaLogDirsRequest(partitionDirs: Map[TopicPartition, String]): AlterReplicaLogDirsResponse = {
-    val request = new AlterReplicaLogDirsRequest.Builder(partitionDirs.asJava).build()
+    val x = partitionDirs.groupBy{case (tp, dir) => dir}.map{ case(dir, tps) =>
```
nit: `x`

tombentley commented:
Rebased for conflict.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -69,6 +69,11 @@
 import org.apache.kafka.common.internals.KafkaFutureImpl;
 import org.apache.kafka.common.message.AlterPartitionReassignmentsRequestData;
 import org.apache.kafka.common.message.AlterPartitionReassignmentsRequestData.ReassignableTopic;
+import org.apache.kafka.common.message.AlterReplicaLogDirsRequestData;
```
Would it make sense to add unit tests in `KafkaAdminClientTest` to cover those changes?

dajac commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -188,8 +188,8 @@ public Struct parseResponse(short version, ByteBuffer buffer) {
-    ALTER_REPLICA_LOG_DIRS(34, "AlterReplicaLogDirs", AlterReplicaLogDirsRequest.schemaVersions(),
-            AlterReplicaLogDirsResponse.schemaVersions()),
+    ALTER_REPLICA_LOG_DIRS(34, "AlterReplicaLogDirs", AlterReplicaLogDirsRequestData.SCHEMAS,
+                           AlterReplicaLogDirsResponseData.SCHEMAS),
```
nit: Alignment is not consistent with the others.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsResponse.java:
```diff
@@ -17,122 +17,59 @@
+        this.data = new AlterReplicaLogDirsResponseData(struct, version);
     }
 
     /**
      * Constructor for version 0.
      */
```
I think that we can removed this comment.

dajac reviewed (commented):
@tombentley Thanks for the PR. I have made a pass over it and I have left couple of comments.

tombentley committed:
Review comments

tombentley committed:
review comments

tombentley committed:
Add test for KafkaAdminClient.alterReplicaLogDirs

tombentley committed:
Add test for KafkaAdminClient.alterReplicaLogDirs

tombentley committed:
Add test for AlterLogDirs RPCs to RequestResponseTests

tombentley committed:
Suggested code by @dajac

tombentley committed:
Add KafkaApisTest.testAlterReplicaLogDirs()

tombentley committed:
Fix test

tombentley commented:
@dajac thanks for the review, I've addressed all your comments.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2207,21 +2212,33 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
             TopicPartitionReplica replica = entry.getKey();
             String logDir = entry.getValue();
             int brokerId = replica.brokerId();
             TopicPartition topicPartition = new TopicPartition(replica.topic(), replica.partition());
```
`topicPartition` is not used except for getting the topic and the partition above. `replica.topic()` and `replica.partition()` could be directly used instead.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2207,21 +2212,33 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
-            if (!replicaAssignmentByBroker.containsKey(brokerId))
-                replicaAssignmentByBroker.put(brokerId, new HashMap<>());
-            replicaAssignmentByBroker.get(brokerId).put(topicPartition, logDir);
+            AlterReplicaLogDirsRequestData value = replicaAssignmentByBroker.computeIfAbsent(brokerId,
```
nit: Could we rename `value` to something like `alterReplicaLogDirs`?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2207,21 +2212,33 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
+                alterReplicaLogDirTopic = new AlterReplicaLogDirTopic();
+                alterReplicaLogDir.topics().add(alterReplicaLogDirTopic);
+            }
+            alterReplicaLogDirTopic.setName(topicPartition.topic())
```
`setName` could be done only once within the if statement.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsResponse.java:
```diff
@@ -17,122 +17,53 @@
-    /**
-     * Possible error code:
-     *
-     * LOG_DIR_NOT_FOUND (57)
-     * KAFKA_STORAGE_ERROR (56)
-     * REPLICA_NOT_AVAILABLE (9)
-     * UNKNOWN (-1)
```
Could we keep those and add them in the javadoc of `AlterReplicaLogDirsResponse` like we did [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java#L31)?

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3388,78 @@ public void testAlterClientQuotas() throws Exception {
+    public void testAlterReplicaLogDirsSuccess() throws Exception {
+        Node node0 = new Node(0, "localhost", 8120);
+        Node node1 = new Node(1, "localhost", 8121);
+        try (AdminClientUnitTestEnv env = mockClientEnv()) {
```
nit: The mocked environment creates 3 nodes (by default) that you can use so you don't have to create them. You can get them with `env.getCluster().nodeById(..)`.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3388,78 @@ public void testAlterClientQuotas() throws Exception {
+        }
+    }
+
+    private void createAlterLogDirsResponse(AdminClientUnitTestEnv env, Node node, Errors error, int... partition) {
```
nit: `partitions`?

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3388,78 @@ public void testAlterClientQuotas() throws Exception {
+            logDirs.put(tpr1, "/data1");
+            AlterReplicaLogDirsResult result = env.adminClient().alterReplicaLogDirs(logDirs);
+            result.values().get(tpr0).get();
+            result.values().get(tpr1).get();
```
nit: We could use `assertNull` here and in the other tests when the future returns `null`.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3388,78 @@ public void testAlterClientQuotas() throws Exception {
+            createAlterLogDirsResponse(env, node1, Errors.NONE, 1, 2);
+
+            TopicPartitionReplica tpr0 = new TopicPartitionReplica("topic", 0, 0);
+            TopicPartitionReplica tpr1 = new TopicPartitionReplica("topic", 0, 1);
```
Do you intentionally use a `TopicPartitionReplica` which is not provided in the response here? I guess that it is to ensure that its future is not completed, isn't it?

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3388,78 @@ public void testAlterClientQuotas() throws Exception {
+            logDirs.put(tpr0, "/data0");
+            logDirs.put(tpr1, "/data1");
+            AlterReplicaLogDirsResult result = env.adminClient().alterReplicaLogDirs(logDirs);
+            // alterReplicaLogDirs() error handling fails all futures, but some of them may already be completed
```
Couldn't we just remove `tpr0` from the test then? It does not seem to bring much.

dajac commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -2198,4 +2203,31 @@ private OffsetDeleteResponse createOffsetDeleteResponse() {
+                        new AlterReplicaLogDirsRequestData.AlterReplicaLogDirTopicCollection(Collections.singletonList(
+                                new AlterReplicaLogDirsRequestData.AlterReplicaLogDirTopic().setPartitions(singletonList(0)).setName("topic")
+                        ).iterator())
+                )
```
nit: Could we break these long lines like you did bellow already?

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2589,15 +2589,27 @@ class KafkaApis(val requestChannel: RequestChannel,
       new DescribeConfigsResponse(requestThrottleMs, (authorizedConfigs ++ unauthorizedConfigs).asJava))
   }
 
+
```
nit: this extra empty line can be removed.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -17,141 +17,84 @@
-        struct.set(LOG_DIRS_KEY_NAME, logDirStructArray.toArray());
-        return struct;
+    @Override
+    public AlterReplicaLogDirsResponse getErrorResponse(Throwable e) {
```
nit: This one is not used if not mistaken. Let's remove it.

dajac commented on core/src/test/scala/unit/kafka/server/AlterReplicaLogDirsRequestTest.scala:
```diff
@@ -104,13 +112,26 @@ class AlterReplicaLogDirsRequestTest extends BaseRequestTest {
+          new AlterReplicaLogDirsRequestData.AlterReplicaLogDirTopic()
+            .setName(topic)
+            .setPartitions(tpPartitions.map{case (tp, _) => tp.partition.asInstanceOf[Integer]}.toList.asJava)
+        }.toList.asJava.iterator))
```
nit: Indentation looks wrong here.

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -1707,4 +1707,38 @@ class KafkaApisTest {
+    EasyMock.reset(replicaManager, clientRequestQuotaManager, requestChannel)
+
+    val capturedResponse = expectNoThrottling()
+    EasyMock.expect(replicaManager.alterReplicaLogDirs(EasyMock.eq(Map(
+      new TopicPartition("t0", 0) -> "/foo",
+      new TopicPartition("t0", 1) -> "/foo",
+      new TopicPartition("t0", 2) -> "/foo"))))
+      .andReturn(Map(
+        new TopicPartition("t0", 0) -> Errors.NONE,
+        new TopicPartition("t0", 1) -> Errors.LOG_DIR_NOT_FOUND,
+        new TopicPartition("t0", 2) -> Errors.INVALID_TOPIC_EXCEPTION))
```
nit: Indentation seems inconsistent.

dajac commented on core/src/test/scala/unit/kafka/server/AlterReplicaLogDirsRequestTest.scala:
```diff
@@ -36,6 +37,13 @@ class AlterReplicaLogDirsRequestTest extends BaseRequestTest {
   val topic = "topic"
 
+  private def findErrorForPartition(response: AlterReplicaLogDirsResponse, tp: TopicPartition): Errors = {
+    Errors.forCode(response.data.results.asScala.groupBy(x => x.topicName).get(tp.topic).get
```
I find the implementation that you did in `AuthorizerIntegrationTest` at L195 easier to read. Could we use it here as well to stay consistent?

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -1707,4 +1707,38 @@ class KafkaApisTest {
+
+    val response = readResponse(ApiKeys.ALTER_REPLICA_LOG_DIRS, alterReplicaLogDirsRequest, capturedResponse)
+      .asInstanceOf[AlterReplicaLogDirsResponse]
+    assertEquals(Map(Errors.NONE -> 1,
```
It is good to validate the `errorCounts` but could we also validate the overall mapping from TopicPartition to Errors?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -17,141 +17,84 @@
-            responseMap.put(entry.getKey(), Errors.forException(e));
-        }
-        return new AlterReplicaLogDirsResponse(throttleTimeMs, responseMap);
+    public AlterReplicaLogDirsResponse getErrorResponse(int throttleTimeMs, Throwable e) {
```
Could we add a unit test for this one and the one bellow?

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsResponse.java:
```diff
@@ -17,122 +17,53 @@
     @Override
     public Map<Errors, Integer> errorCounts() {
-        return errorCounts(responses.values());
+        Map<Errors, Integer> errorCounts = new HashMap<>();
```
Could we add a unit test for this one?

dajac reviewed (commented):
@tombentley Thanks for the updated PR. The overall logic looks good to me. I have left a bunch of small comments.

tombentley committed:
review comments

tombentley commented:
Thanks @dajac, there were some useful comments and good spots there. I've pushed some fixes.

dajac commented on clients/src/test/java/org/apache/kafka/common/requests/AlterReplicaLogDirsResponseTest.java:
```diff
@@ -0,0 +1,57 @@
+public class AlterReplicaLogDirsResponseTest {
+
+    @Test
+    public void testErrorResponse() {
```
I suppose that this should be named `testErrorCounts`.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3388,70 @@ public void testAlterClientQuotas() throws Exception {
+            TopicPartitionReplica tpr1 = new TopicPartitionReplica("topic", 0, 1);
+
+            Map<TopicPartitionReplica, String> logDirs = new HashMap<>();
+            assertNull(logDirs.put(tpr0, "/data0"));
+            assertNull(logDirs.put(tpr1, "/data1"));
```
`assertNull`s shouldn't be here but few lines bellow.

tombentley replied:
Good catch! Sorry about that.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2234,18 +2249,18 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
+                            KafkaFutureImpl<Void> future = futures.get(replica);
+                            if (future == null) {
+                                handleFailure(new IllegalStateException(
+                                        "The partition " + new TopicPartition(topicResult.topicName(), partitionResult.partitionIndex()) + " in the response from broker " + brokerId + " is not in the request"));
```
@tombentley This one bugs me. I know that it was already like this in the previous implementation but I wonder if we could improve this while we are here. I am also fine keeping it as it is and improving this latter on.

I find odd that we fail all the non-realised yet futures when we discover an unknown replica in the response. From a client perspective, the behaviour is unpredictable and we leave it with potentially a partial view of the results event though the brokers may have sent back all the requested ones. Wouldn't it be better to just ignore unexpected replica and log a warning message? If a broker would ever do this, the client would continue to work as expected which is probably better. I think that it is a better behavior.

Similarly, we may want to ensure that all futures are done in case a replica would be missing in the response from the broker.

What do you think?

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -1707,4 +1707,47 @@ class KafkaApisTest {
+      .asInstanceOf[AlterReplicaLogDirsResponse]
+    assertEquals(partitionResults, response.data.results.asScala.flatMap { tr =>
+      tr.partitions().asScala.map { pr =>
+        new TopicPartition(tr.topicName(), pr.partitionIndex()) -> Errors.forCode(pr.errorCode())
```
nit: `()` can be removed for all the accessors here.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequest.java:
```diff
@@ -17,141 +17,79 @@
 
     public Map<TopicPartition, String> partitionDirs() {
-        return partitionDirs;
+        Map<TopicPartition, String> result = new HashMap<>();
```
Do you mind adding a unit test for this one in `AlterReplicaLogDirsRequestTest` for the sake of completeness?

dajac reviewed (commented):
@tombentley Thanks for the updated PR. It looks good overall. I have left few more small comments and one question.

tombentley committed:
Review comments; change handling of unexpected responses

tombentley commented:
Rebased for conflict.

@dajac thanks for the review. I agree about the case of the unexpected partition. The only way it could happen would be a bug in either the client or the broker, but there's no need to force that onto the user to handle. A warning log should be enough for the bug to be reported eventually. So I've made this change.

About the case of the client not getting a response for a partition: I agree that could happen (again, as a bug in the broker or client). It would be a bit harder to detect though; we'd be spending CPU cycles to detect and log a case which we don't think will ever happen in practice. Maybe that's OK since altering log dirs not something done frequently. But it strikes me that there are plenty of other RPCs in the admin client that don't do this kind of second-guessing of the results. If we're prepared to go to that effort here, why not elsewhere? So personally, I think it would OK to omit this check, but I can do it if you really think it should be done.

tombentley committed:
Review comments 2

dajac commented:
@tombentley Did you consider adding a loop in at the end of the `handleResponse` that would complete any non-completed future related to the current `brokerId` with an exception?

```
// The server should send back a response for every topic. But do a sanity check anyway.
for (Map.Entry<TopicPartitionReplica, KafkaFutureImpl<Void>> entry : futures.entrySet()) {
    TopicPartitionReplica replica = entry.getKey();
    KafkaFutureImpl<Void> future = entry.getValue();
    if (!future.isDone && replica.brokerId() == brokerId)) {
        future.completeExceptionally(new ApiException("The server response did not " +
            "contain a reference to node " + entry.getKey()));
    }
}
```

You're right. We have few other cases in the Admin client that suffer from the same issue. If we can find an easy way to do it, I would rather prefer doing it in this PR so we can forget about it. I will try to fix some of the others myself.

tombentley committed:
Exception on partial response

tombentley commented:
@dajac yes, fair point. And I notice now that create and delete topics response handlers are already doing this (though the error messages in those cases don't really make much sense). So I've added this with a slightly reworded message.

dajac commented:
@tombentley thanks for the addition. we could perhaps add a small unit test for the sake of completeness. we already cover all the other cases but this one.

tombentley committed:
Add extra test

tombentley committed:
Two more tests

tombentley commented:
@dajac sure. I added tests for the create and delete topics cases too.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -477,6 +484,27 @@ public void testCreateTopics() throws Exception {
+                    prepareCreateTopicsResponse("myTopic", Errors.NONE));
+            CreateTopicsResult topicsResult = env.adminClient().createTopics(
+                    asList(new NewTopic("myTopic", Collections.singletonMap(0, asList(0, 1, 2))),
+                            new NewTopic("myTopic2", Collections.singletonMap(0, asList(0, 1, 2)))),
```
nit: Can we align it with the `new` above?

dajac commented on clients/src/test/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequestTest.java:
```diff
@@ -0,0 +1,88 @@
+                        asList(new AlterReplicaLogDir()
+                                .setPath("/data0")
+                                .setTopics(new AlterReplicaLogDirTopicCollection(
+                                        asList(new AlterReplicaLogDirTopic()
+                                                .setName("topic")
+                                                .setPartitions(asList(0, 1)),
+                                                new AlterReplicaLogDirTopic()
+                                                        .setName("topic2")
+                                                        .setPartitions(asList(7))).iterator())),
```
nit: indentation is a bit inconsistent here.

dajac commented on clients/src/test/java/org/apache/kafka/common/requests/AlterReplicaLogDirsRequestTest.java:
```diff
@@ -0,0 +1,88 @@
+                                        .setPath("/data1")
+                                        .setTopics(new AlterReplicaLogDirTopicCollection(
+                                                asList(new AlterReplicaLogDirTopic()
+                                                                .setName("topic3")
+                                                                .setPartitions(asList(12))).iterator()))).iterator()));
```
nit: Indentation of these two lines look weird.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2234,18 +2258,29 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
+                    for (Map.Entry<TopicPartitionReplica, KafkaFutureImpl<Void>> entry : futures.entrySet()) {
+                        TopicPartitionReplica replica = entry.getKey();
+                        KafkaFutureImpl<Void> future = entry.getValue();
+                        if (!future.isDone()
+                                && replica.brokerId() == brokerId) {
```
nit: i would bring the second part on the first line.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1402,6 +1407,15 @@ int numPendingCalls() {
+    /**
+     * Fail the given future when a response handler expected a result for an entity but no result was present.
+     * @param future The future to fail.
+     * @param message The message to fail the future with
```
nit: `.` at the end to stay consistent with previous line.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1402,6 +1407,15 @@ int numPendingCalls() {
+     * @param future The future to fail.
+     * @param message The message to fail the future with
+     */
+    private void partialResponse(KafkaFutureImpl<?> future, String message) {
```
I am not entirely convinced by the name. At the end, the method complete a future with the given message. There is nothing specific to handling partial responses. I would rename it to something more generic or simply remove it as we don't gain much with it.

tombentley replied:
I agree it's not doing much. What value it has is in trying to handle these cases in a consistent way, and being able to more easily discover/reason about the call sites. Maybe something like `invalidBrokerResponse()` would be a better name? But if you don't like that I'm happy to remove it.

dajac replied:
I see your point. I am definitely for reusing logic as much as possible for those cases. What about going an extra step and do something like this then (code not tested)? Just an idea...

```
public static <K, V> void completeUnrealizedFutures(
            Map<K, KafkaFutureImpl<V>> futures,
            Function<K, Boolean> filter,
            Function<K, String> messageFormatter) {
        for (Map.Entry<K, KafkaFutureImpl<V>> entry : futures.entrySet()) {
            K key = entry.getKey();
            KafkaFutureImpl<V> future = entry.getValue();
            if (!future.isDone() && filter.apply(key)) {
                future.completeExceptionally(new ApiException(messageFormatter.apply(key)));
            }
        }
    }

    public static  <K, V> void completeUnrealizedFutures(
            Map<K, KafkaFutureImpl<V>> futures,
            unction<K, String> messageFormatter) {
        completeUnrealizedFutures(futures, (key) -> true, messageFormatter);
    }
```

It would allow the following:

```
completeUnrealizedFutures(topicFutures,
     replica -> replica.brokerId() == brokerId,
     topic -> "The response from broker " + brokerId + " did not contain a result for replica " + replica);

completeUnrealizedFutures(topicFutures,
     topic -> "The server response did not contain a reference to node " + topic);
```

tombentley replied:
That works for me, especially if this is something you're going to make more use of for the other response handlers in the admin client.

dajac reviewed (commented):
@tombentley Thanks for the update! LGTM. I just left small cosmetic comments.

tombentley committed:
Code review

tombentley committed:
completeUnrealizedFutures

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1479,13 +1512,9 @@ public void handleResponse(AbstractResponse abstractResponse) {
-                }
+                completeUnrealizedFutures(topicFutures,
+                    topic -> "The controller response " +
+                            "did not contain a result for topic " + topic);
```
nit: we can bring this back on the first line.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1552,13 +1581,9 @@ void handleResponse(AbstractResponse abstractResponse) {
-                }
+                completeUnrealizedFutures(topicFutures,
+                    topic -> "The controller response " +
+                            "did not contain a result for topic " + topic);
```
nit: we can bring this back on the first line.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2234,20 +2269,26 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
-                            future.completeExceptionally(error.exception());
+                    for (AlterReplicaLogDirTopicResult topicResult: response.data().results()) {
+                        for (AlterReplicaLogDirPartitionResult partitionResult: topicResult.partitions()) {
+                            TopicPartitionReplica replica = new TopicPartitionReplica(topicResult.topicName(), partitionResult.partitionIndex(), brokerId);
```
nit: We could break this long line.

dajac reviewed (approved):
@tombentley Thanks for the update. The PR LGTM! I just left few cosmetic comments.

tombentley committed:
Review comments + improvement

tombentley commented:
@dajac I fixed those, and also changed the `completeUnrealizedFutures()` method to take a `Stream`, which is a little nicer since it allows the caller to do the filtering, thus removing the need for the overloaded version.

dajac commented:
@tombentley thanks!

hachikuji commented:
ok to test

mimaison commented:
retest this please

mimaison commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -477,6 +484,27 @@ public void testCreateTopics() throws Exception {
+            try {
+                topicsResult.values().get("myTopic2").get();
+                fail("Expected an exception.");
+            } catch (ExecutionException e) {
```
We can use `assertThrows()` here. Same below

tombentley replied:
I don't think we can, at least not if we want to check that `topicsResult.values().get("myTopic2").get()` throws an `ExecutionException` _wrapping an `ApiException`_. `assertThrows()` would only let us assert the outer exception type.

mimaison replied:
Sorry, I meant `TestUtils.assertFutureThrows()`

tombentley replied:
Ah, thanks! Now fixed.

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2234,20 +2251,27 @@ public AlterReplicaLogDirsResult alterReplicaLogDirs(Map<TopicPartitionReplica,
+                                    topicResult.topicName(), partitionResult.partitionIndex(), brokerId);
+                            KafkaFutureImpl<Void> future = futures.get(replica);
+                            if (future == null) {
+                                log.warn("The partition {} in the response from broker {}} is not in the request",
```
There's an extra `}`

mimaison commented:
Thanks @tombentley for the PR. It looks good. Happy to merge once the last 2 comments are addressed

tombentley committed:
review comment

tombentley committed:
TestUtils.assertFutureThrows

mimaison commented:
retest this please

mimaison commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -570,6 +592,27 @@ public void testDeleteTopics() throws Exception {
+                    new DeleteTopicsOptions()).values();
+            values.get("myTopic").get();
+
+            try {
```
We can use `TestUtils.assertFutureThrows()` here too

mimaison commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -3382,6 +3425,90 @@ public void testAlterClientQuotas() throws Exception {
+            logDirs.put(tpr2, "/data1");
+            AlterReplicaLogDirsResult result = env.adminClient().alterReplicaLogDirs(logDirs);
+            assertNull(result.values().get(tpr1).get());
+            try {
```
we can use `TestUtils.assertFutureThrows()` here too

tombentley committed:
More TestUtils.assertFutureThrows

tombentley commented:
@mimaison done.

mimaison commented:
retest this please
